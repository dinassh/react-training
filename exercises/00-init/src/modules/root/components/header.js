import React from 'react'
import PropTypes from 'prop-types'

// Exercise 1: implement Header component
// export class Header extends React.Component
// {

//     static propTypes = {
//         title: PropTypes.string.isRequired
//     }

//     render() {
//         return (
//             <h1>{this.props.title}</h1>
//         )
//     }
// }





// Exercise 2: implement Header as function
export const Header = (props) =>
{
    return <h1>{props.title}</h1>
}

Header.propTypes = {
    title: PropTypes.string.isRequired
}