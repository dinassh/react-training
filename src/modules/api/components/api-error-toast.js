import React from "react";
import { connect } from "react-redux";

import { getError } from "modules/api/api-selectors";

const ApiErrorToast = ({ error }) => error && <div>{error}</div>;

const mapStateToProps = state => ({
  error: getError(state)
});

export default connect(mapStateToProps)(ApiErrorToast);
