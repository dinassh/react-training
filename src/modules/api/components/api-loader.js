import React from "react";
import { connect } from "react-redux";

import { isLoading } from "modules/api/api-selectors";

export const ApiLoader = ({ loading }) => loading && <div>Loading...</div>;

const mapStateToProps = state => ({
  loading: isLoading(state)
});

export default connect(mapStateToProps)(ApiLoader);
