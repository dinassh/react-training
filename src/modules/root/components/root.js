import React from "react";
import { Portal } from "react-portal";

import ApiErrorToast from "modules/api/components/api-error-toast";
import ApiLoader from "modules/api/components/api-loader";
import Route from "modules/routing/components/route";
import UsersRoute from "modules/users/components/users-route";

const Root = () => (
  <div>
    <h1>This is a testing application</h1>
    <Route startsWith="users" component={UsersRoute} />
    <Portal>
      <ApiErrorToast />
      <ApiLoader />
    </Portal>
  </div>
);

export default Root;
