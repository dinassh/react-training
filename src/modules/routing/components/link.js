import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { actions } from "redux-router5";

const Link = ({ onClick, children }) => <a onClick={onClick}>{children}</a>;

Link.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired
};

const mapDispatchToProps = {
  onClick: actions.navigateTo
};

export default connect(
  null,
  mapDispatchToProps,
  (stateProps, dispatchProps, ownProps) => ({
    ...ownProps,
    onClick: () => dispatchProps.onClick(ownProps.name, ownProps.params)
  })
)(Link);
