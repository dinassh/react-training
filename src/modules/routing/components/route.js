import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { startsWithSegment, endsWithSegment } from "router5-helpers";

import { getCurrentRoute } from "modules/routing/routing-selectors";

export const Route = ({
  route: { name },
  startsWith,
  endsWith,
  exact,
  component: Component
}) => {
  if (startsWith && !endsWith && !exact) {
    return startsWithSegment(name)(startsWith) && <Component />;
  } else if (endsWith && !startsWith && !exact) {
    return endsWithSegment(name)(endsWith) && <Component />;
  } else if (exact && !startsWith && !endsWith) {
    return exact === name && <Component />;
  } else {
    throw new Error(
      "Invalid state, you need to either provide startsWith, endsWith or exact prop"
    );
  }
};

Route.propTypes = {
  route: PropTypes.shape({
    name: PropTypes.string.isRequired
  }).isRequired,
  startsWith: PropTypes.string,
  endsWith: PropTypes.string,
  exact: PropTypes.string,
  component: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  route: getCurrentRoute(state)
});

export default connect(mapStateToProps)(Route);
