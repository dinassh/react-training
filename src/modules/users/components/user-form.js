import React from "react";
import PropTypes from "prop-types";
import { reduxForm, FieldArray } from "redux-form";
import { connect } from "react-redux";

import FormField from "modules/forms/components/form-field";
import FormFieldSelect from "modules/forms/components/form-field-select";
import * as Validations from "modules/forms/validations";
import { getSkills } from "modules/crud/crud-selectors";

export const USER_FORM_NAME = "user-form";

const validateSkillNotEmpty = Validations.notEmptyObject("Skill is required");
const validateNotEmptyFirstName = Validations.notEmptyString(
  "First Name is required"
);
const validateNotEmptyLastName = Validations.notEmptyString(
  "Last Name is required"
);
const validateEmailNotEmpty = Validations.notEmptyString("E-mail is required");

const Skills = ({ skills, fields, meta: { error, submitFailed, dirty } }) => (
  <div>
    <h2>Skills</h2>
    {(submitFailed || dirty) && error && <div>{error}</div>}
    <ul>
      {fields.map((name, index) => (
        <li key={index}>
          <FormFieldSelect
            name={`${name}.skill`}
            options={skills}
            valueKey="id"
            labelKey="skill"
            validate={validateSkillNotEmpty}
          />
          <button type="button" onClick={() => fields.remove(index)}>
            x
          </button>
        </li>
      ))}
    </ul>
    <button type="button" onClick={() => fields.push({})}>
      Add skill
    </button>
  </div>
);

Skills.propTypes = {
  skills: PropTypes.array.isRequired,
  fields: PropTypes.object.isRequired,
  meta: PropTypes.shape({
    error: PropTypes.string,
    submitFailed: PropTypes.bool,
    dirty: PropTypes.bool
  }).isRequired
};

const UserForm = ({ handleSubmit, skills, onSubmit }) => (
  <form onSubmit={handleSubmit(onSubmit)}>
    <label>First Name</label>
    <FormField
      name="firstName"
      type="text"
      placeholder="John"
      validate={validateNotEmptyFirstName}
    />
    <label>Last Name</label>
    <FormField
      name="lastName"
      type="text"
      placeholder="Doe"
      validate={validateNotEmptyLastName}
    />
    <label>E-mail</label>
    <FormField
      name="email"
      type="text"
      placeholder="email@email.com"
      validate={[validateEmailNotEmpty, Validations.isEmail]}
    />
    <FieldArray
      name="skills"
      component={Skills}
      skills={skills}
      validate={[Validations.notEmptySkills, Validations.allSkillsUnique]}
    />
    <button>Save</button>
  </form>
);

UserForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  skills: PropTypes.array.isRequired
};

const mapStateToProps = state => ({
  skills: getSkills(state)
});

export default reduxForm({
  form: USER_FORM_NAME,
  enableReinitialize: true
})(connect(mapStateToProps)(UserForm));
