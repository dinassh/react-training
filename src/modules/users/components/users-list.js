import React from "react";

const UsersList = () => <div>Users List</div>;
export default UsersList;
