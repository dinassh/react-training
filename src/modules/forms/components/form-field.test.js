import React from "react";
import renderer from "react-test-renderer";

import { FieldAdapter } from "modules/forms/components/form-field";

describe("Form Field Tests", () => {
  it("should not render warning/error when field has not been touched", () => {
    const cmp = renderer.create(
      <FieldAdapter
        input={{}}
        label="foobar"
        type="text"
        meta={{
          touched: false,
          error: "some error",
          warning: "some warning"
        }}
      />
    );

    expect(cmp.toJSON()).toMatchSnapshot();
  });

  it("should render warning/error when field has been touched", () => {
    const cmp = renderer.create(
      <FieldAdapter
        input={{}}
        label="foobar"
        type="text"
        meta={{
          touched: true,
          error: "some error"
        }}
      />
    );

    expect(cmp.toJSON()).toMatchSnapshot();
  });
});
