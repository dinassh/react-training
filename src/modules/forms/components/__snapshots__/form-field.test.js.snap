// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Field Tests should not render warning/error when field has not been touched 1`] = `
<div>
  <label>
    foobar
  </label>
  <div>
    <input
      placeholder="foobar"
      type="text"
    />
  </div>
</div>
`;

exports[`Form Field Tests should render warning/error when field has been touched 1`] = `
<div>
  <label>
    foobar
  </label>
  <div>
    <input
      placeholder="foobar"
      type="text"
    />
    <span>
      some error
    </span>
  </div>
</div>
`;
